core/src/main/java/org/apache/accumulo/core/clientImpl/ClientContext.java

```diff
@@ -164,13 +175,48 @@ public ClientContext(SingletonReservation reservation, ClientInfo info,
     this.singletonReservation = Objects.requireNonNull(reservation);
     this.tableops = new TableOperationsImpl(this);
     this.namespaceops = new NamespaceOperationsImpl(this, tableops);
+    String uehClassName = serverConf.get(ClientProperty.UNCAUGHT_EXCEPTION_HANDLER.getKey());
```

Comment:
The way this is being read from the config is a bit weird. It's reading a ClientProperty from an AccumuloConfiguration, which means that the client property would have had to exist in the `accumulo.properties` file, but not documented as a server-side property. The property is explicitly named `client.uncaught.exception.handler`, but the way it is loaded, it will set it for the server-side as well. I think this would be cleaner if it were renamed to `instance.uncaught.exception.handler` and put in `Property.java` with other server properties. You could even use ClassLoaderUtil directly, or via ConfigurationTypeHelper to load the class from the configuration.

If you wanted a separate client-side property name, you could do that (see how ClientConfConverter translates ZooKeeper host names, for an example), but I don't think that'd be necessary.

Reply:
I think I'm misunderstanding something as I'm not familiar with the new Accumulo client in 2.x. Does the Accumulo client in 2.x use the `accumulo-client.properties` file? I was thinking that it did, now I'm thinking that it does not. Is that file only used for the shell and utilities?

Reply:
The client does use that file, but the server-side doesn't. This code is shared by both client and server code because `ClientContext` is the `Impl` class for `AccumuloClient`, and it is the class that is returned from the `AccumuloClient` builder. `ClientContext` is also the super-class of `ServerContext`, because servers can also perform client operations.

The builder for `AccumuloClient` can initialize the configuration used in `ClientContext` from the client configuration file. However, the `ServerContext` initializes the configuration used in `ClientContext` from the server's `conf/accumulo.properties` file (which clients cannot be expected to have access to). Complicating the situation, we have two different versions of client properties, and have a translator tool to convert one into the other. And, to share code between the two use cases, we present the client configuration file to `ClientContext` in an `AccumuloConfiguration` view.

Because this class is the super-class of `ServerContext`, we would need it to be able to read the value from the `conf/accumulo.properties` file for the server-side. Users can have the option of putting the same configuration property in their own `accumulo-client.properties` file, if they wish (and that's the main use case of this PR). For other properties where that happens, we try to keep the same name, so it's easier. However, for historical reasons, some of the properties, like the one for setting the ZooKeeper hosts, have a different property name that we translate if we see it.

Yeah, it's complicated... but that's the trade-off for preserving deprecated functionality, properties, key names, etc. I would *love* to strip out the old stuff, and dramatically simplify a lot of this, but that has been resisted in the past (understandably). I'm hoping we can clean it up with a non-LTM 3.0 release before we start working on an LTM 3.x version.

Reply:
@ctubbsii - Thanks for the explanation, that helped. I'm about to push a commit that:

1. addresses your comments above about lazy initialization of the thread pools
2. removes the UEH handler property entirely
3. Sets the UEH to `Threads.UEH` when `ClientContext` is initialized from `ServerContext`
4. Allows the client to set a UEH class on the AccumuloClient when building it
5. Uses a default UEH in the client if one is not supplied that logs the error. This should provide some backwards compatibility for clients so the client VM does not get terminated.